unit/goto-symex/symex_assign.cpp

```diff
@@ -192,4 +192,61 @@ SCENARIO(
       }
     }
   }
+  GIVEN(
+    "A symbol `struct1` a with_exprt `struct1 with [field1 <- 234]` and "
+    "a skeleton `☐.field_1`")
```

Comment:
This should NOT have been unit tested, but rather be covered in a regression test. The unit test merely makes sure that the code does whatever the author thinks it does, instead of testing what the user should expect. The trace output is all about providing value to humans and other actual users. But obviously I'm too late to the party.

Reply:
Sometimes the function has an obvious bug but it is difficult to hit the problematic behavior with a regression test, and even if we can hit that behavior now, it may no longer be the case when other parts of the code change. In those cases unit-test are the right thing to add because it is easier to hit the piece of code you want, and it is tested independently of the rest.

Reply:
Unit tests can be useful, and the point about making sure it is hit is a good one. But I don't accept "difficult" as a reason not to do other testing. (I typically spend most of my time designing tests rather than coding the fix.) Via unit tests we may as well be testing code that is entirely irrelevant and should possibly be removed. If future changes result in this code no longer being covered by regression tests then surely something is going wrong: either the result is no longer as expected, or the code has become redundant (and should be removed).

Reply:
In this particular case I'm adding a new class and this unit-test can act as documentation by example of what the class is doing, and in addition there is value in testing that the code does what the author thinks it does.
Unit tests for a user of this function which is `assign_symbol` are also included in this PR, so we are also testing that the user is using it right.

Reply:
I'm sorry, I was ambiguous in my use of the term "user." I care about the end user, i.e., human running CBMC. These are the only "users" that matter to me.

Reply:
:+1: but I also care about developers and they read these unit tests (sometimes)